shakedown/dcos/__init__.py

```diff
@@ -162,3 +260,86 @@ def _gen_url(url_path):
     """
     from six.moves import urllib
     return urllib.parse.urljoin(dcos_url(), url_path)
+
+
+# Initialize some things.
+transportManager = TransportManager()
+
+
+class HostSession:
+    """Context manager that returns an SSH session to run commands.
+    
+    """
+    def __init__(self, host, username, key_path, verbose):
+        self.host = host
+        self.username = username
+        self.key_path = key_path
+        self.verbose = verbose
+        self.exit_code = -1
+        self.output = ''
+        self.session = None
+    
+    def __enter__(self):
+        """Called when using a `with` closure.
+
+        :return: this session manager
+        :rtype: HostSession
+        """
+        self.session = transportManager.get_session(
+                self.host,
+                self.username,
+                self.key_path)
+        return self
+    
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        """Executed when the context manager is complete (exits
+        the with closure).
+
+        :param exc_type:
+        :param exc_val:
+        :param exc_tb:
+        :return: None
+        """
+        self.exit_code = self.session.recv_exit_status()
+        self._wait_for_recv()
+        # read data that is ready
+        while self.session.recv_ready():
+            rl, wl, xl = select.select([self.session], [], [], 0.0)
+            if len(rl) > 0:
+                recv = str(self.session.recv(1024), "utf-8")
+                if self.verbose:
+                    print(recv, end='', flush=True)
+                self.output += recv
+        try_close(self.session)
+        return None
+    
+    def _wait_for_recv(self):
+        """After executing a command, wait for results.
+        
+        Because `recv_ready()` can return False, but still have a
+        valid, open connection, it is not enough to ensure output
+        from a command execution is properly captured.
+
+        :return: None
+        """
+        event = threading.Event()
+        while True:
+            event.wait(0.1)
+            if self.session.recv_ready():
+                return
+            if self.session.closed:
```

Comment:
couldn't this just be an `or` condition?

Reply:
Yes. I kept them separate because they are different scenarios, but not opposed to combining them.